File: algorithms/02_c/data_structures/segment_tree.c
#include <stdio.h>

#define MAXN 100000

static long long tree[4 * MAXN];

void seg_build(const long long *a, int node, int l, int r) {
    if (l == r) { tree[node] = a[l]; return; }
    int m = (l + r) / 2;
    seg_build(a, node * 2, l, m);
    seg_build(a, node * 2 + 1, m + 1, r);
    tree[node] = tree[node * 2] + tree[node * 2 + 1];
}

void seg_update(int node, int l, int r, int idx, long long val) {
    if (l == r) { tree[node] = val; return; }
    int m = (l + r) / 2;
    if (idx <= m) seg_update(node * 2, l, m, idx, val);
    else seg_update(node * 2 + 1, m + 1, r, idx, val);
    tree[node] = tree[node * 2] + tree[node * 2 + 1];
}

long long seg_query(int node, int l, int r, int ql, int qr) {
    if (ql > r || qr < l) return 0;
    if (ql <= l && r <= qr) return tree[node];
    int m = (l + r) / 2;
    return seg_query(node * 2, l, m, ql, qr) + seg_query(node * 2 + 1, m + 1, r, ql, qr);
}

int main(void) {
    long long a[] = {1, 2, 3, 4, 5};
    seg_build(a, 1, 0, 4);

    if (seg_query(1, 0, 4, 0, 4) != 15 || seg_query(1, 0, 4, 1, 3) != 9) {
        printf("[C SegmentTree] FAILED: initial query\n");
        return 1;
    }
    seg_update(1, 0, 4, 2, 10);
    if (seg_query(1, 0, 4, 0, 4) != 22 || seg_query(1, 0, 4, 2, 2) != 10) {
        printf("[C SegmentTree] FAILED: post-update query\n");
        return 1;
    }
    printf("[C SegmentTree] Point update + range sum verified\n");
    return 0;
}

File: algorithms/07_typescript/data_structures/segment_tree.ts
export class SegmentTree {
  private n: number;
  private tree: number[];

  constructor(arr: number[]) {
    this.n = arr.length;
    this.tree = new Array(4 * this.n).fill(0);
    this.build(arr, 1, 0, this.n - 1);
  }

  private build(arr: number[], node: number, l: number, r: number): void {
    if (l === r) { this.tree[node] = arr[l]; return; }
    const mid = (l + r) >> 1;
    this.build(arr, node * 2, l, mid);
    this.build(arr, node * 2 + 1, mid + 1, r);
    this.tree[node] = this.tree[node * 2] + this.tree[node * 2 + 1];
  }

  update(pos: number, val: number, node: number = 1, l: number = 0, r: number = this.n - 1): void {
    if (l === r) { this.tree[node] = val; return; }
    const mid = (l + r) >> 1;
    if (pos <= mid) this.update(pos, val, node * 2, l, mid);
    else this.update(pos, val, node * 2 + 1, mid + 1, r);
    this.tree[node] = this.tree[node * 2] + this.tree[node * 2 + 1];
  }

  query(qL: number, qR: number, node: number = 1, l: number = 0, r: number = this.n - 1): number {
    if (qL > r || qR < l) return 0;
    if (qL <= l && r <= qR) return this.tree[node];
    const mid = (l + r) >> 1;
    return this.query(qL, qR, node * 2, l, mid) + this.query(qL, qR, node * 2 + 1, mid + 1, r);
  }
}

const st = new SegmentTree([1, 3, 5, 7, 9, 11]);
if (st.query(1, 3) !== 15) throw new Error("Segment Tree failed");
st.update(2, 10);
if (st.query(1, 3) !== 20) throw new Error("Segment Tree update failed");
console.log("[TypeScript Segment Tree] Range sum with point update verified");
